File: CSharp/Oomph.Data/UF09Test/UFs/v401/ABC277_C.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Oomph.Data.UF09Lib.UFs.v401;

namespace UF09Test.UFs.v401
{
	// Test: https://atcoder.jp/contests/abc277/tasks/abc277_c
	class ABC277_C
	{
		static int[] Read() => Array.ConvertAll(Console.ReadLine().Split(), int.Parse);
		static (int u, int v) Read2() { var a = Read(); return (a[0], a[1]); }
		static void Main() => Console.WriteLine(Solve());
		static object Solve()
		{
			var n = int.Parse(Console.ReadLine());
			var es = Array.ConvertAll(new bool[n], _ => Read2());

			var keys = es.Select(e => e.u).Concat(es.Select(e => e.v)).Append(1);
			var uf = new UnionFind<int>(keys);
			foreach (var (a, b) in es)
			{
				uf.Union(a, b);
			}
			return uf.ToGroups()[uf.Find(1)].Max();
		}
	}
}

File: Python/oomph-data-ws/uf-309-lib/ABC277-C.py
# https://atcoder.jp/contests/abc277/tasks/abc277_c

from unionfind506 import UnionFindWithData

n = int(input())
edges = [tuple(int(s) for s in input().split()) for _ in range(n)]

# compression
set = set()
set.add(1)
for a, b in edges:
    set.add(a)
    set.add(b)
f = [x for x in set]
f.sort()
m = len(f)

map = {}
for i in range(m):
    map[f[i]] = i

# max2 = lambda x, y: x if x >= y else y
uf = UnionFindWithData(f, max)
for a, b in edges:
    uf.union(map[a], map[b])
print(uf[0])
